packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts

```diff
@@ -0,0 +1,425 @@
+  public static readonly Ec2 = new VpcEndpointAwsService('ec2');
+  public static readonly Ec2Messages = new VpcEndpointAwsService('ec2messages');
+  public static readonly Ecr = new VpcEndpointAwsService('ecr.api');
+  public static readonly EcrDocker = new VpcEndpointAwsService('ecr.dkr');
+  public static readonly Ecs = new VpcEndpointAwsService('ecs');
+  public static readonly EcsAgent = new VpcEndpointAwsService('ecs-agent');
+  public static readonly EcsTelemetry = new VpcEndpointAwsService('ecs-telemetry');
+  public static readonly ElasticInferenceRuntime = new VpcEndpointAwsService('elastic-inference.runtime');
+  public static readonly ElasticLoadBalancing = new VpcEndpointAwsService('elasticloadbalancing');
+  public static readonly CloudWatchEvents = new VpcEndpointAwsService('events');
+  public static readonly ApiGateway = new VpcEndpointAwsService('execute-api');
+  public static readonly CodeCommitGit = new VpcEndpointAwsService('git-codecommit');
+  public static readonly CodeCommitGitFips = new VpcEndpointAwsService('git-codecommit-fips');
+  public static readonly KinesisStreams = new VpcEndpointAwsService('kinesis-streams');
+  public static readonly Kms = new VpcEndpointAwsService('kms');
+  public static readonly CloudWatchLogs = new VpcEndpointAwsService('logs');
+  public static readonly CloudWatch = new VpcEndpointAwsService('monitoring');
+  public static readonly S3 = new VpcEndpointAwsService('s3', VpcEndpointType.Gateway);
+  public static readonly SageMakerApi = new VpcEndpointAwsService('sagemaker.api');
+  public static readonly SageMakerRuntime = new VpcEndpointAwsService('sagemaker.runtime');
+  public static readonly SageMakerRuntimeFips = new VpcEndpointAwsService('sagemaker.runtime-fips');
+  public static readonly SecretsManager = new VpcEndpointAwsService('secretsmanager');
+  public static readonly ServiceCatalog = new VpcEndpointAwsService('servicecatalog');
+  public static readonly Sns = new VpcEndpointAwsService('sns');
+  public static readonly Sqs = new VpcEndpointAwsService('sqs');
+  public static readonly Ssm = new VpcEndpointAwsService('ssm');
+  public static readonly SsmMessages = new VpcEndpointAwsService('ssmmessages');
+  public static readonly Sts = new VpcEndpointAwsService('sts');
+  public static readonly Transfer = new VpcEndpointAwsService('transfer.server');
+
+  constructor(name: string, type?: VpcEndpointType, prefix?: string) {
+    super(`${prefix || 'com.amazonaws'}.${cdk.Aws.region}.${name}`, type || VpcEndpointType.Interface);
+  }
+}
+
+/**
+ * Options to add an endpoint to a VPC.
+ */
+export interface VpcEndpointOptions {
+  /**
+   * The name of the service.
+   */
+  readonly service: VpcEndpointService;
+}
+
+/**
+ * Options to add a gateway endpoint to a VPC.
+ */
+export interface VpcGatewayEndpointOptions extends VpcEndpointOptions {
+  /**
+   * Where to add endpoint routing.
+   *
+   * @default private subnets
+   */
+  readonly subnets?: SubnetSelection[]
```

Comment:
Does this really need to be an array?

Reply:
For a VPC with isolated and private subnets, the user could want to add endpoint routing to the route tables of both subnet types.

Reply:
See the same kind of discussion here for VPN routing https://github.com/awslabs/aws-cdk/pull/1899#discussion_r260963485.

If possible, I would like to refactor here to also use a `SubnetSelection[]`:
https://github.com/awslabs/aws-cdk/blob/7001f7723cde611f0d8cd9c182f92120b09579c3/packages/%40aws-cdk/aws-ec2/lib/vpc.ts#L141-L147

Reply:
I wonder if it wouldn't be a better approach to make the SubnetSelection able to represent "private + isolated", don't you think?

Reply:
You mean changing the `SubnetSelection` interface to:
```ts
export interface SubnetSelection {
  readonly subnetType?: SubnetType[];

  readonly subnetName?: string[];
}
```

We also have to support a list of `subnetName`s if the user created the `VpcNetwork` with a specific `subnetConfiguration`

Reply:
+1 on extending SubnetSelection. That's our go-to